### ISSUE TITLE
Exception leak

### ISSUE THREAD

**Issue by matusfedorko** [ThePhD/sol2#1183]
As far as I understand, sol should be able to translate C++ exceptions thrown from code bound to Lua via sol. The code below works in Visual Studio in `Debug` build, but fails in `Release` mode. In `Release` , I get `Unhandled exception at ...`.

I am not sure if this is the MSVC compiler to blame or sol library. And also I saw several exception related bugs have already been reported , so feel free to close this issue if it is a duplicate. As far as I can tell, I am using latest sol (checked out develop branch) and Visual Studio 2019.
```
#include <sol/sol.hpp>
#include <iostream>

static void funcThrow()
{
	throw std::runtime_error("Whoopsie");
}

int main()
{
	std::cerr << "================= script exception leak ===================" << std::endl;

	sol::state lua;
	lua.open_libraries(sol::lib::base, sol::lib::package, sol::lib::coroutine, sol::lib::string, sol::lib::os, sol::lib::math, sol::lib::table, sol::lib::io, sol::lib::debug);

	// leaks exception in release
	lua["funcThrow"] = [f = &funcThrow] { return f(); };

	// works ok
	//lua["funcThrow"] = &funcThrow;

	// also works
	//lua["funcThrow"] = [] { return funcThrow(); };

	auto res = lua.safe_script("funcThrow()", sol::script_pass_on_error);
	if (!res.valid())
		std::cerr << "status=" << int(res.status()) << ", error=" << res.get<sol::error>().what() << std::endl;

	std::cerr << "================= THE END ===================" << std::endl;

	return 0;
}
```

**Comment by matusfedorko** [comment]
Another update:

```
lua["funcThrow"] = sol::overload(std::function<void()>([f = &funcThrow] { return f(); }));
```

still leaks the exception.

**Comment by matusfedorko** [comment]
As a temporary workaround, I have changed `struct unqualified_pusher<overload_set<Functions...>>::push` to call `function_detail::select_set_fx<false, true, F>` instead of `function_detail::select_set_fx<false, false, F>`.
This selects `detail::static_trampoline<function_detail::call<meta::unqualified_t<Fx>, 2, is_yielding>>` instead of `function_detail::call<meta::unqualified_t<Fx>, 2, is_yielding>`, which contains catch statements to handle exceptions, the latter one produces codepath with no catch clauses.

**Comment by ThePhD** [comment]
This is part of the ongoing saga "please unbreak exceptions, they are borked", which has like 6 issues open in its name. I'm planning to fix all of these relatively soon, but I've had to re-work my testing and CMake lately so that building and maintaining all of this is not such a nightmare.

**Comment by matusfedorko** [comment]
Thanks for being so active in maintaining this awesome library. Great work !

**Comment by ThePhD** [comment]
Minor question; do you have any other defines in here? I'm compiling this and all of them are not throwing whatsoever.

**Comment by matusfedorko** [comment]
No, not as far, as I can tell. Are you running `Release` build ? Because in Debug it works somehow. And I forgot to mention, I am running 64-bit version, not x86, maybe that is significant too.

I have zipped up my entire test project. It is slightly different than what I posted here (should not be anything significant though). And you will have to setup correct include and library directories for your machine, but I hope it helps.
[exception.zip](https://github.com/ThePhD/sol2/files/6430264/exception.zip)
